### PR TITLE
CIFAR benchmark 

### PR DESCRIPTION
**Pull request by Landanjs**
CIFAR benchmark.

I couldn't figure out how to do tests with synthetic local data since the torchvision CIFAR10 dataset is very particular. I was able to setup a local MDS dataset that can be used for testing. Although, I'm not sure if we should have streaming in this benchmark.

I copied as much as possible form composer. We didn't have previous recipes (I think?), so I added something random. 

I didn't add the dataset check as in the other benchmarks since it is fairly easy to download the data using the torchvision dataset. I also didn't include multi-node information since I didn't think that made sense for CIFAR10.

WandB results: 93.25 baseline; 93.92 recipe (one seed each

Any ideas for a figure? I can do wandb plots, but need to figure out dark mode like Abhi 😅

### CODE REVIEW

**Review by A-Jacobson — APPROVED** [review]
I think it's fine to just have runnable code for CIFAR. The benefit of this benchmark (to me) is that it's quick and easy to run. I doubt people care much for specific recipes on cifar, it's so quick and easy you'd likely have a weird combination of methods overfitting the eval process anyway.

File: cifar/model.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+def build_composer_resnet_cifar(model_name: str, num_classes = 10):
+
+    def weight_init(w: torch.nn.Module):
+        if isinstance(w, nn.Linear) or isinstance(w, nn.Conv2d):
+            torch.nn.init.kaiming_normal_(w.weight)
+        if isinstance(w, nn.BatchNorm2d):
+            w.weight.data = torch.rand(w.weight.data.shape)
+            w.bias.data = torch.zeros_like(w.bias.data)
+
+
```
Comment (dblalock) [inline cifar/model.py:130]:
```suggestion

```

File: cifar/yamls/resnet56.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
+run_name: cifar10_base # Name of the  training run used for checkpointing and other logging
+is_train: true         # Trains the model if true, otherwise runs evaluation
+seed: 42               # Random seed
+max_duration: 160ep    # Duration to train specified as a Time string
+use_recipe: false      # Whether to train with a recipe for faster convergence or higher accuracy
+grad_accum: 'auto'     # Amount of gradient accumulation, 'auto' (GPU only) means Composer will set the optimal value
+
+train_dataset:
+  path: cifar10
+  is_streaming: false # Set to `true` if your data is in a remote location (e.g. a S3 bucket)
```
Comment (dblalock) [inline cifar/yamls/resnet56.yaml:10]:
```suggestion
  is_streaming: false # Set to `true` if your data is in a remote location (e.g. an S3 bucket)
```
also in the other `is_streaming:` line below

**Review by dblalock — APPROVED** [review]
LGTM; just some minor local stuff

File: cifar/model.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            self.relu = nn.ReLU(inplace=True)
+
+            # No parameters for shortcut connections.
+            if downsample or f_in != f_out:
+                self.shortcut = nn.Sequential(
+                    nn.Conv2d(f_in, f_out, kernel_size=1, stride=2, bias=False),
+                    nn.BatchNorm2d(f_out),
+                )
+            else:
+                self.shortcut = nn.Sequential()
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply cifar/model.py:44]: I took this from composer, probably so?

File: cifar/model.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    @staticmethod
+    def is_valid_model_name(model_name: str):
+        valid_model_names = [f'resnet_{layers}' for layers in (20, 56)]
+        return (model_name in valid_model_names)
+
+    @staticmethod
+    def get_model_from_name(model_name: str, initializer: Optional[Callable] = None, num_classes: int = 10):
+        """The naming scheme for a ResNet is ``'resnet_D[_W]'``.
+        D is the model depth (e.g. ``'resnet_56'``)
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply cifar/model.py:97]: I can delete the W, I'm not sure where the cifar resnet implementation came from

File: cifar/data.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                                 shuffle=shuffle,
+                                 transform=transform,
+                                 batch_size=batch_size)
+        sampler = None
+    else:
+        with dist.run_local_rank_zero_first():
+            dataset = datasets.CIFAR10(root=data_path,
+                                       train=is_train,
+                                       download=dist.get_local_rank() == 0 and download,
+                                       transform=transform)
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply cifar/data.py:136]: The run_local_rank_zero_first calls barrier on all process that are not rank zero and calls barrier on the rank zero process once it is done with the context, so I think it is okay? I haven't run into any issue after several runs

File: cifar/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+[epoch=0][batch=17/625]: memory/alloc_retries: 3
+[epoch=0][batch=17/625]: trainer/grad_accum: 2
+[epoch=0][batch=17/625]: loss/train/total: 7.1243
+[epoch=0][batch=17/625]: metrics/train/Accuracy: 0.0010
+train          Epoch   0:    3%|▋                        | 17/625 [00:17<07:23,  1.37ba/s, loss/train/total=7.1292]
+```
+
+# Saving and Loading checkpoints
+
+At the bottom of `yamls/resnet50.yaml`, we provide arguments for saving and loading model weights. Please specify the `save_folder` or `load_path` arguments if you need to save or load checkpoints!
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply cifar/README.md:117]: Ooops, imagenet was on my mind. I didn't specify defaults since this would automatically save/load a checkpoint. I figured we wouldn't this be the default behavior in case the user doesn't want to save checkpoints yet
